pmd-core/src/main/java/net/sourceforge/pmd/lang/document/TextFileContent.java

```diff
@@ -0,0 +1,323 @@
+    }
+
+    /**
+     * Read the reader fully and produce a {@link TextFileContent}. This
+     * closes the reader. This takes care of buffering.
+     *
+     * @param reader A reader
+     *
+     * @return A text file content
+     *
+     * @throws IOException If an IO exception occurs
+     */
+    public static TextFileContent fromReader(Reader reader) throws IOException {
+        try (Reader r = reader) {
+            return normalizingRead(r, DEFAULT_BUFSIZE, FALLBACK_LINESEP, newChecksum(), true);
+        }
+    }
+
+
+    /**
+     * Reads the contents of the input stream into a TextFileContent.
+     * This closes the input stream. This takes care of buffering.
+     *
+     * @param inputStream    Input stream
+     * @param sourceEncoding Encoding to use to read from the data source
+     */
+    public static TextFileContent fromInputStream(InputStream inputStream, Charset sourceEncoding) throws IOException {
+        return fromInputStream(inputStream, sourceEncoding, FALLBACK_LINESEP);
+    }
+
+    // test only
+    static TextFileContent fromInputStream(InputStream inputStream, Charset sourceEncoding, String fallbackLineSep) throws IOException {
+        Checksum checksum = newChecksum();
+        try (CheckedInputStream checkedIs = new CheckedInputStream(new BufferedInputStream(inputStream), checksum);
+             // no need to buffer this reader as we already use our own char buffer
+             Reader reader = new InputStreamReader(checkedIs, sourceEncoding)) {
+            return normalizingRead(reader, DEFAULT_BUFSIZE, fallbackLineSep, checksum, false);
+        }
+    }
+
+    // test only
+    static @NonNull TextFileContent normalizeCharSeq(CharSequence text, String fallBackLineSep) {
+        long checksum = getCheckSum(text); // the checksum is computed on the original file
+
+        if (text.length() > 0 && text.charAt(0) == ByteOrderMark.UTF_BOM) {
+            text = text.subSequence(1, text.length()); // skip the BOM
+        }
+        Matcher matcher = NEWLINE_PATTERN.matcher(text);
+        boolean needsNormalization = false;
+        String lineTerminator = null;
+        while (matcher.find()) {
+            lineTerminator = detectLineTerm(lineTerminator, matcher.group(), fallBackLineSep);
+            if (!NORMALIZED_LINE_TERM.equals(lineTerminator)) {
+                needsNormalization = true;
+
+                if (lineTerminator.equals(fallBackLineSep)) {
+                    // otherwise a mixed delimiter may follow, and we must
+                    // detect it to fallback on the system separator
+                    break;
+                }
+            }
+        }
+        if (lineTerminator == null) {
+            // no line sep, default to platform sep
+            lineTerminator = fallBackLineSep;
+            needsNormalization = false;
+        }
+
+        if (needsNormalization) {
+            text = NEWLINE_PATTERN.matcher(text).replaceAll(NORMALIZED_LINE_TERM);
```

Comment:
Hm... if we normalize e.g. `\r\n` into `\n`, then we change the text... Addressing content by linenumber + column should still be ok, but addressing by offset will be different. E.g. if we report the location of a violation, it might be rendered at a different position by tools because we analyzed not the original text, but a normalized text...
Won't this normalization create more problems than it solves?

Reply:
Offsets are not output anywhere, they're valid on a given text document in memory, but are not absolute (eg with document views, the same place may have different offsets depending on whether you ask the view or its parent). We still only report violations with line/column. So I don't think that's a problem.

Reply:
> only report violations with line/column

I think, that's key. It's maybe something to keep in mind - PMD won't be able to support a renderer that wants to output offset coordinates. I've checked the SARIF format - there both coordinate systems are supported, so we are fine here.